### PR TITLE
Normalize ambiguous characters in captcha

### PR DESCRIPTION
**Pull request by frederikrothenberger**
This PR makes the captcha easier to solve for humans by normalizing confusing characters (such as `l` and `1`).

E.g. this means that no captcha containing l will be generated and all inputs into captcha solutions will map `l` to `1`.

### CODE REVIEW

File: src/internet_identity/src/anchor_management/registration.rs [hunk 1]
```diff
@@ -10,9 +10,11 @@ use ic_cdk::{call, caller, trap};
 use internet_identity_interface::archive::{DeviceDataWithoutAlias, Operation};
 use internet_identity_interface::*;
 use rand_core::{RngCore, SeedableRng};
+use std::collections::HashMap;
 
 #[cfg(not(feature = "dummy_captcha"))]
-use captcha::filters::Wave;
+use captcha::filters::{Grid, Wave};
```
Comment (nmattia) [inline src/internet_identity/src/anchor_management/registration.rs:16]: is `Grid` needed?

File: src/internet_identity/src/anchor_management/registration.rs [hunk 2]
```diff
@@ -139,10 +160,22 @@ fn create_captcha<T: RngCore>(rng: T) -> (Base64, String) {
 
 // Check whether the CAPTCHA challenge was solved
 fn check_challenge(res: ChallengeAttempt) -> Result<(), ()> {
+    // avoid processing too many characters
+    if res.chars.len() > CAPTCHA_LENGTH {
+        return Err(());
+    }
```
Comment (nmattia) [inline src/internet_identity/src/anchor_management/registration.rs:166]: This is unrelated, right?

File: src/internet_identity/src/anchor_management/registration.rs [hunk 3]
```diff
@@ -118,11 +120,30 @@ fn create_captcha<T: RngCore>(rng: T) -> (Base64, String) {
...
 #[cfg(not(feature = "dummy_captcha"))]
 fn create_captcha<T: RngCore>(rng: T) -> (Base64, String) {
     let mut captcha = captcha::RngCaptcha::from_rng(rng);
+    let mut chars = captcha.supported_chars();
+    chars.retain(|c| !CHAR_REPLACEMENTS.contains_key(c));
+
     let captcha = captcha
-        .add_chars(5)
+        .set_chars(&chars)
+        .add_chars(CAPTCHA_LENGTH as u32)
```
Comment (nmattia) [inline src/internet_identity/src/anchor_management/registration.rs:146]: Would it be simpler to normalize the challenge result instead?

File: src/internet_identity/src/anchor_management/registration.rs (hunk 2 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/anchor_management/registration.rs:166]: I added it since we now do processing on the input where previously we did only an equality check.

File: src/internet_identity/src/anchor_management/registration.rs (hunk 3 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/anchor_management/registration.rs:146]: The challenge result is normalized. But I can also have the captcha generated _with_ ambiguous characters and normalize both, the challenge and the input by the user when checking the result. Is this what you would prefer?

File: src/internet_identity/src/anchor_management/registration.rs (hunk 1 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/anchor_management/registration.rs]: Ah, no. Removed.

File: src/internet_identity/src/anchor_management/registration.rs (hunk 3 above)
Comment (nmattia) [reply src/internet_identity/src/anchor_management/registration.rs:146]: Ah my bad, I was confused by `CAPTCHA_LENGTH`. I thought you were adding as many chars to the captcha as there were in `CHAR_REPLACEMENTS` and then only got more confused from there.

If the challenge result is normalized, why also replace `chars`? Is that just for consistency?

File: src/internet_identity/src/anchor_management/registration.rs [hunk 4]
```diff
@@ -139,10 +160,22 @@ fn create_captcha<T: RngCore>(rng: T) -> (Base64, String) {
 
 // Check whether the CAPTCHA challenge was solved
 fn check_challenge(res: ChallengeAttempt) -> Result<(), ()> {
+    // avoid processing too many characters
+    if res.chars.len() > CAPTCHA_LENGTH {
+        return Err(());
+    }
+    // Normalize challenge attempts by replacing characters that are not in the captcha character set
+    // with the respective replacement from CHAR_REPLACEMENTS.
+    let normalized_challenge: String = res
```
Comment (nmattia) [inline src/internet_identity/src/anchor_management/registration.rs:169]:
```suggestion
    let normalized_challenge_res: String = res
```

got confused and thought you were re-normalizing the challenge

File: src/internet_identity/src/anchor_management/registration.rs (hunk 3 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/anchor_management/registration.rs:146]: I think we have 2 options:
1. Normalize (reduce) input chars, such that the captcha does not contain ambiguous chars (i.e. `1lXcx` will never be generated) and also normalize the input by the user.
2. Generate Captcha as before (i.e. `1lXcx`) and only when checking the solution normalize _both_ the actual challenge and the user input. So in this example the user input would be normalize and checked against `11xcx`.

This PR so far implements approach 1.

**Review by nmattia — APPROVED** [review]
LGTM modulo "challenge" vs "challenge result" (see suggestion)

File: src/internet_identity/src/anchor_management/registration.rs [hunk 5]
```diff
@@ -118,11 +120,30 @@ fn create_captcha<T: RngCore>(rng: T) -> (Base64, String) {
     return (resp, captcha.chars_as_string());
 }
 
+lazy_static! {
+    /// Map of problematic characters that are easily mixed up by humans to the "normalized" replacement.
+    /// I.e. the captcha will never contain any of the characters in the key set and any input provided
+    /// will be mapped to the matching value.
+    /// Note: the captcha library already excludes the characters o, O and 0.
```
[the remark this answers is not in the document]
Comment (frederikrothenberger) [reply src/internet_identity/src/anchor_management/registration.rs:127]: No, I suspect it is because for those characters it is less controversial that they are ambiguous. Contributing more to the library could also help us make use of more filters.

Happy to discuss offline.
